A shard chain eventual consistency implementation

leventeliu opened the pull request:
- Add xenomint -- a shard chain eventual consistency implementation
- Refactor kayak by @xq262144 
- Redesign sqlchain, use kayak+xenomint to improve performance
- Fix a lot of issues

leventeliu committed:
Add the xenomint storage interface definition

leventeliu committed:
Add the sqlite implementation of xenomint storage

leventeliu committed:
Add basic type definitions of xenomint

leventeliu committed:
Add xenomint package doc.go

leventeliu committed:
Remove forgotten sleeps which are replaced by ticker

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Add rlimit setting in testing setup

leventeliu committed:
Fix idle interval to 1 ms

leventeliu committed:
Also remove share memory and WAL files

leventeliu committed:
Fix reader retry interval to 1 ms

leventeliu committed:
Add timer ops in benchmarking

leventeliu committed:
Rename some testing functions

leventeliu committed:
Add testing initialization

leventeliu committed:
Add chain state definition

leventeliu committed:
Add testing for chain state

leventeliu committed:
Add tx pool definition

leventeliu committed:
Add errors definition

leventeliu committed:
Add benchmarking for read over transaction

leventeliu committed:
Add state replaying support

leventeliu committed:
Add benchmarking for sequential write

leventeliu committed:
Add partial commit

leventeliu committed:
Add benchmarking for sequential transaction write

leventeliu committed:
Add basic query interface and benchmarking

leventeliu committed:
Fix error assertion

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Add mux service benchmarking

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Add bp service in benchmark setup return values

leventeliu committed:
Add RPC mux service and benchmarking

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,98 @@
+	serviceMap sync.Map
+}
+
+func NewMuxService(name string, server *rpc.Server) (service *MuxService, err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function NewMuxService should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,98 @@
+	s.serviceMap.Delete(id)
+}
+
+type MuxQueryRequest struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxQueryRequest should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,98 @@
+	Request *wt.Request
+}
+
+type MuxQueryResponse struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxQueryResponse should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,98 @@
+	return
+}
+
+func (s *MuxService) Query(req *MuxQueryRequest, resp *MuxQueryResponse) (err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method MuxService.Query should have comment or be unexported

leventeliu committed:
Use a 333 column length to approach 1 Kbyte record

leventeliu committed:
Add key space setting

leventeliu committed:
Wrap locking scope to prevent deadlock

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

laodouya committed:
Refactor remove extra setup in bench sqlite test.

laodouya committed:
Add setup progress info bench sqlite test.

leventeliu committed:
Add runtime trace support

leventeliu committed:
Use common setup to speed up benchmarking

leventeliu committed:
Comment out runtime tarcing

leventeliu committed:
Add basic chain runtime

covenantdb commented on xenomint/chain.go:
```diff
@@ -0,0 +1,265 @@
+	}
+}
+
+func (c *Chain) Stop() (err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method Chain.Stop should have comment or be unexported

covenantdb commented on xenomint/chain.go:
```diff
@@ -0,0 +1,265 @@
+	return c.state.close(true)
+}
+
+func (c *Chain) Start() (err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method Chain.Start should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	return
+}
+
+type MuxApplyRequest struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxApplyRequest should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	Response *wt.Response
+}
+
+type MuxApplyResponse struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxApplyResponse should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	return
+}
+
+type MuxLeaderCommitRequest struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxLeaderCommitRequest should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	Height int32
+}
+
+type MuxLeaderCommitResponse struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxLeaderCommitResponse should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	Offset uint64
+}
+
+// Commit is the RPC method to commit block on mux service.
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported method MuxService.LeaderCommit should be of the form "LeaderCommit ..."

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	return
+}
+
+type MuxFollowerCommitRequest struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxFollowerCommitRequest should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	Offset uint64
+}
+
+type MuxFollowerCommitResponse struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MuxFollowerCommitResponse should have comment or be unexported

covenantdb commented on xenomint/mux.go:
```diff
@@ -0,0 +1,193 @@
+	proto.Envelope
+}
+
+func (s *MuxService) FollowerCommit(
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method MuxService.FollowerCommit should have comment or be unexported

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Remove old serialization code

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Use crypto.HSV suite to sign and verify block

covenantdb commented on sqlchain/types/block.go:
```diff
@@ -40,83 +40,38 @@ type Header struct {
+	HSV crypto.DefaultHashSignVerifierImpl
+}
+
+func (s *SignedHeader) Sign(signer *ca.PrivateKey) error {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method SignedHeader.Sign should have comment or be unexported

leventeliu committed:
Keep old packages for hotfix

leventeliu committed:
Fix broken dependencies

Commit:
More verbose log for kayak runtime

leventeliu committed:
Move sqlchain and worker types to top level

This change will resolve the import cycle between sqlchain/types and worker/types.

covenantdb commented on types/block.go:
```diff
@@ -0,0 +1,151 @@
+	HSV crypto.DefaultHashSignVerifierImpl
+}
+
+func (s *SignedHeader) Sign(signer *ca.PrivateKey) error {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method SignedHeader.Sign should have comment or be unexported

leventeliu committed:
Fix broken type conversion

Commit:
Support command line args in docker-entry for miner/blockproducer

leventeliu committed:
Remove obsolete types

leventeliu committed:
Remove redundant imports

leventeliu committed:
Renew generated codes

Commit:
Temporary disable leveldb buffer setting, use default 8MiB

leventeliu committed:
Add Query and Replay methods for sqlchain

covenantdb commented on types/block.go:
```diff
@@ -0,0 +1,158 @@
+	return s.Verify()
+}
+
+type QueryAsTx struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type QueryAsTx should have comment or be unexported

covenantdb commented on xenomint/pool.go:
```diff
@@ -0,0 +1,54 @@
+	"github.com/CovenantSQL/CovenantSQL/types"
+)
+
+type QueryTracker struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type QueryTracker should have comment or be unexported

covenantdb commented on xenomint/pool.go:
```diff
@@ -0,0 +1,54 @@
+	resp *types.Response
+}
+
+func (q *QueryTracker) UpdateResp(resp *types.Response) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method QueryTracker.UpdateResp should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,404 @@
+	"github.com/pkg/errors"
+)
+
+type State struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type State should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,404 @@
+	return
+}
+
+func (s *State) Query(req *types.Request) (ref *QueryTracker, resp *types.Response, err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.Query should have comment or be unexported

Commit:
Refactor non-chain types to use DefaultHashSignVerifierImpl

Commit:
Move blockproducer database service types to types package

Commit:
Use verifier.DefaultHashSignVerifierImpl in blockproducer types

Commit:
Unify sign/verification errors

leventeliu committed:
Add method to commit block on xenomint state

leventeliu committed:
Fix issue on committing

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,494 @@
+	return
+}
+
+func (s *State) ReplayBlock(block *types.Block) (err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.ReplayBlock should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,494 @@
+	return
+}
+
+func (s *State) CommitEx() (queries []*QueryTracker, err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.CommitEx should have comment or be unexported

leventeliu committed:
Add block producing logic

covenantdb commented on xenomint/pool.go:
```diff
@@ -0,0 +1,108 @@
+	q.Resp = resp
+}
+
+func (q *QueryTracker) Ready() bool {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method QueryTracker.Ready should have comment or be unexported

covenantdb replied:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method QueryTracker.Ready should have comment or be unexported

leventeliu committed:
Switch block producing to V2

leventeliu committed:
Add comments for GoDoc

covenantdb commented on xenomint/pool.go:
```diff
@@ -0,0 +1,108 @@
+	"github.com/CovenantSQL/CovenantSQL/types"
+)
+
+type QueryTracker struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type QueryTracker should have comment or be unexported

covenantdb commented on xenomint/pool.go:
```diff
@@ -0,0 +1,108 @@
+	Resp *types.Response
+}
+
+func (q *QueryTracker) UpdateResp(resp *types.Response) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method QueryTracker.UpdateResp should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,502 @@
+	"github.com/pkg/errors"
+)
+
+type State struct {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type State should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,502 @@
+	current uint64 // current is the current savepoint of the current transaction
+}
+
+func NewState(strg xi.Storage) (s *State, err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function NewState should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,502 @@
+	s.current = id
+}
+
+func (s *State) InitTx(id int32) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.InitTx should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,502 @@
+	return
+}
+
+func (s *State) ReplayBlock(block *types.Block) (err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.ReplayBlock should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,502 @@
+	return
+}
+
+func (s *State) CommitEx() (queries []*QueryTracker, err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.CommitEx should have comment or be unexported

covenantdb commented on xenomint/state.go:
```diff
@@ -0,0 +1,502 @@
+	return
+}
+
+func (s *State) Query(req *types.Request) (ref *QueryTracker, resp *types.Response, err error) {
```
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.Query should have comment or be unexported

leventeliu committed:
Add comments for GoDoc

leventeliu committed:
Add unit testing for xenomint state

Commit:
Golint issues

leventeliu committed:
Fix state transaction leak on closing

leventeliu committed:
Add unit testing for state replay

leventeliu committed:
Fix query type error on testing

leventeliu committed:
Add more test cases

leventeliu committed:
Fix commit id issue and remove transaction id

leventeliu committed:
Fix a state conflict case

Commit:
Integration without kayak sqlchain kayak wal

Commit:
Compatible with current sqlparser mysql dialect sql syntax

Commit:
Add leveldb datastore to .gitignore list

Commit:
Reduce build stage docker image size

leventeliu committed:
Remove unused codes

leventeliu committed:
Add map to keep track of failed request

Commit:
Update observer structure

leventeliu committed:
Add method to calculate state id for reloading

Commit:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Add ack index and fix block producing

leventeliu committed:
Merge remote-tracking branch 'origin/develop' into feature/consistency

leventeliu committed:
Switch to go1.10.x and close runtime trace

leventeliu committed:
Fix observer compiling error caused by struct change

leventeliu committed:
Minor fix for observer

leventeliu committed:
Add test cases and remove unused codes

leventeliu committed:
Fix route testing

auxten commented on sqlchain/observer.go:
```diff
@@ -21,23 +21,17 @@ import (
-	wt "github.com/CovenantSQL/CovenantSQL/worker/types"
 )
 
 /*
```
why leave comment code here?

xq262144 replied:
The comment here is a brief introduction of api requirements for a simple observer service.

auxten commented on xenomint/state.go:
```diff
@@ -0,0 +1,617 @@
+	// Check if the current transaction is ok to commit
+	if s.pool.matchLast(lastsp) {
+		if err = s.uncCommit(); err != nil {
+			// FATAL ERROR
```
why not just log.FATAL

leventeliu replied:
It actually needs some further processing here, or it may run into an unrecoverable state. And usually, a Tx without context doesn't produce error on committing. Maybe leave a TODO here for later optimization.

auxten reviewed (approved):
LGTM
